core/src/main/scala/kafka/coordinator/transaction/TransactionMarkerRequestCompletionHandler.scala

```diff
@@ -22,25 +22,25 @@ import kafka.utils.Logging
 import org.apache.kafka.clients.{ClientResponse, RequestCompletionHandler}
 import org.apache.kafka.common.TopicPartition
 import org.apache.kafka.common.protocol.Errors
-import org.apache.kafka.common.requests.WriteTxnMarkersRequest.TxnMarkerEntry
 import org.apache.kafka.common.requests.WriteTxnMarkersResponse
 
 import scala.collection.mutable
 import collection.JavaConversions._
 
-class TransactionMarkerRequestCompletionHandler(transactionMarkerChannel: TransactionMarkerChannel,
-                                                txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker],
+class TransactionMarkerRequestCompletionHandler(brokerId: Int,
                                                 txnTopicPartition: Int,
-                                                txnMarkerEntries: java.util.List[TxnMarkerEntry],
-                                                brokerId: Int) extends RequestCompletionHandler with Logging {
+                                                txnStateManager: TransactionStateManager,
+                                                txnMarkerChannel: TransactionMarkerChannel,
+                                                txnIdAndMarkerEntries: java.util.List[TxnIdAndMarkerEntry],
+                                                txnMarkerPurgatory: DelayedOperationPurgatory[DelayedTxnMarker]) extends RequestCompletionHandler with Logging {
   override def onComplete(response: ClientResponse): Unit = {
     val correlationId = response.requestHeader.correlationId
     if (response.wasDisconnected) {
       trace(s"Cancelled request $response due to node ${response.destination} being disconnected")
       // re-enqueue the markers
-      for (txnMarker: TxnMarkerEntry <- txnMarkerEntries) {
-        transactionMarkerChannel.addRequestToSend(
-          txnTopicPartition,
+      for (txnIdAndMarker: TxnIdAndMarkerEntry <- txnIdAndMarkerEntries) {
```

Comment:
Hmm, in the disconnected case, shouldn't we check txnStateManager.getTransactionState as well in case an emmigration has happened?